File: C++/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
 public:
  bool isValidBST(TreeNode *root, TreeNode *min = NULL, TreeNode *max = NULL) {
    if (!root) return true;
    if (min != NULL && root->val <= min->val) return false;
    if (max != NULL && root->val >= max->val) return false;
    return isValidBST(root->left, min, root) &&
           isValidBST(root->right, root, max);
  }
};

File: Python/0098-validate-binary-search-tree.py
# time complexixty: O(n)
# space complexity: O(n)
import math
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def helper(root: Optional[TreeNode], low=-math.inf, high=math.inf):
            if not root:
                return True
            if root.val <= low:
                return False
            if root.val >= high:
                return False
            return helper(root.left, low, root.val) and helper(root.right, root.val, high)

        return helper(root)
